src/target/cortexa.c

```diff
@@ -199,6 +240,90 @@ static const char tdesc_cortex_a[] =
 	"    <reg name=\"d15\" bitsize=\"64\" type=\"float\"/>"
 	"  </feature>"
 	"</target>";
+*/
+// Returns the amount of characters written to the buffer.
+static size_t create_tdesc_cortex_a(char *buffer, size_t max_len)
+{
+	size_t total = 0;
+
+	// We can't just repeatedly pass max_len to snprintf, because we keep changing the start
+	// of buffer (effectively changing its size), so we have to repeatedly compute the size
+	// passed to snprintf by subtracting the current total from max_len.
+	// ...Unless max_len is 0, in which case that subtraction will result in an (underflowed)
+	// negative number. So we also have to repeatedly check if max_len is 0 before performing
+	// that subtraction.
+	size_t printsz = max_len;
+
+
+	if (buffer != NULL)
+		memset(buffer, 0, max_len);
+
+	// Start with the "preamble", which is generic across ARM targets,
+	// ...save for one word, so we'll have to do the preamble in halves, and then we'll
+	// follow it with the GDB ARM Core feature tag.
+	total += snprintf(buffer, printsz,
+		"%s feature %s "
+		"<feature name=\"org.gnu.gdb.arm.core\">",
+		gdb_arm_preamble_first,
+		gdb_arm_preamble_second
+	);
+
+	// Then the general purpose registers, which have names of r0 to r12.
+	for (uint8_t i = 0; i <= 12; ++i) {
```

Comment:
we recommend using size_t here as the codegen won't be more expensive but it avoids integer promotion interactions

Reply:
`size_t` won't really avoid integer promotion interactions as variadic arguments are promoted to (signed) `int`. If we want to avoid any actual problems due to integer promotion interactions, then we should either use an unsigned "small" type like `unsigned short` or `uint8_t` to avoid any actual consequences of integer promotion; or we should use a (signed) `int` to prevent promotion in the first place.

That being said, we also already know there cannot be any consequences of integer promotion interactions in this case as the value of `i` is very clearly and strictly bounded.

We're fine with changing it from `uint8_t` as a matter of style, but we don't think that `size_t` specifically makes much sense here.

Reply:
We were referring to promotion shenanigans within the loop, not specifically the sprintf(). Trying to keep the code base additions `-Wconversion` and friends clean and this may ping a warning when we enable that - size_t or uint32_t won't (they're also the same underlying type on ARM which is why we suggested size_t here).

We've also found the transform we're requesting consistently reduces code size by a couple of instructions (strb/ldrb vs str/ldr).

Reply:
So, `strb`/`ldrb` have the same size as `str`/`ldr`, so that's not going to reduce code size. We checked `-Wconversion` and it doesn't raise any warnings there, because `-Wconversion` is for cases where integer promotion changes the value/result, and in the loop itself there aren't any interactions where it changes the result (since it's just simple compares and additions with statically bounded values to literals). Checking `-Wconversion` *did* point out that `snprintf()` returns an `int` and thus adding its return to a `size_t` changes the sign, so for correctness most of the `size_t`s that are in these functions should actually be `int`s.

We could also use `uint32_t` like you suggested (or probably `int32_t`, due to the aforementioned sign issue), but from a correctness standpoint that's equally wrong to `size_t`, and all of these should simply be `int`. `uint32_t` does happen to be typedef'd to `unsigned long` on our platform, which also happens to be the same width as `unsigned int` on our platform, but if we're looking from the standpoint of what Actually Happens instead of from a Correctness standpoint, then none of this matters in the first place — we checked the disassembly and confirmed that none of these changes make any difference.

Reply:
We've talked about this over on Discord and decided a way forward, marking this as resolved pending the update discussed